File: Ex01/ivo_xtof/wordcount.rb
#!/usr/bin/ruby
# Last changed Time-stamp: <2008-03-25 22:25:18 xtof>
# $Id$
#
# synopsis: counts word frequencies; reads from stdin, writes to stdout
#
# usage:
# lynx -dump http://www.tbi.univie.ac.at/Origin/origin_1.html \
# | ruby wordcount.rb                                         \
# | less
#

# create an empty hash, initialize default value to 0
freqs = Hash.new(0)

# read line-by-line from stdin
while line = gets
 # split line at non-alphabetic characters and filter out non-wordes
 words = line.chomp.split(/[^a-zA-Z]/).grep(/[a-zA-z]+/)
 # canonize words to lowercase and store them in hash
 words.each{|word| freqs[word.downcase] += 1}
end

# sort hash keys by frequency and output key-value pairs
freqs.sort_by{|x,y| y}.reverse.each{|word,freq| puts freq.to_s+' '+word}

File: Ex01/ivo_xtof/wordcount.pl
#!/usr/bin/perl
# Last changed Time-stamp: <2008-04-03 21:58:07 xtof>
# $Id$
#
# synopsis: counts word frequencies; reads from stdin, writes to stdout
#
# usage:
# lynx -dump http://www.tbi.univie.ac.at/Origin/origin_1.html \
# | perl wordcount.pl                                         \
# | less
#

use strict;
use vars qw/%WC/;

# read line-by-line from stdin
while (<>) {
  # remove \n and canonicalize to lower case
  chomp; tr/A-Z/a-z/;
  # split at non-alphabetic characters filter out non-words and memorize words
  $WC{$_}++ for grep {length $_} split /[^a-zA-Z]/;
}

# sort hash keys by frequency and output key-value pairs
printf "%2d %s\n", $WC{$_}, $_  for sort {$WC{$b} <=> $WC{$a}
					  || $a cmp $b } keys %WC;
